File: vol1/ch12/LeetCode308.cpp
class SegTree {             //线段树类
public: 
    vector<int> ST;
    vector<int> a;
    int n;
    SegTree() {}
    SegTree(vector<int> &a) {
        n=a.size();
        this->a=a;
        ST.resize(4*n);
        Create(1,1,n);
    }
    void Create(int root,int L,int R) {
        if (L==R)
            ST[root]=a[L-1];
        else {
            int mid=(L+R)>>1;
            Create(2*root,L,mid);
            Create(2*root+1,mid+1,R);
            ST[root]=ST[2*root]+ST[2*root+1];
        }
    }
    void UpdateOne(int root,int L,int R,int i,int x) {          //单点修改:a[i]+=x
        if(L==R) ST[root]=x;                                            //单点: 直接修改
        else {                                          //非单点
            int mid=(L+R)>>1;
            if(i<=mid) UpdateOne(2*root,L,mid,i,x);		//在左子树中修改
            else UpdateOne(2*root+1,mid+1,R,i,x);   //在右子树中修改
            ST[root]=ST[2*root]+ST[2*root+1];       //向上更新分支结点值
        }
    }
    int QueryRange(int root,int L,int R,int i,int j) {          //查询区间[i,j]
        if(L>=i && R<=j) return ST[root];                              //当前结点[L,R]包含在[i,j]中
        else {
            int mid=(L+R)>>1;                           //其他情况
            int lsum=0,rsum=0;
            if(i<=mid) lsum=QueryRange(2*root,L,mid,i,j);     	 	//递归查询左子树
            if(mid<j) rsum=QueryRange(2*root+1,mid+1,R,i,j); 	//递归查询右子树
            return lsum+rsum;
        }
    }
};
class NumMatrix {
public:
    int m,n;
    SegTree st;
    NumMatrix(vector<vector<int>>& matrix) {
        m=matrix.size(),n=matrix[0].size();
        vector<int> a;
        for (int i=0;i<m;i++) {
            for (int j=0;j<n;j++) {
                a.emplace_back(matrix[i][j]);
            }
        }
        st=SegTree(a);
    }
    int pno(int i,int j) {      //二维转换为一维
        return i*n+j+1;
    }
    void update(int r,int c,int val) {
        int i=pno(r,c);
        st.UpdateOne(1,1,m*n,i,val);
    }
    int sumRegion(int r1,int c1,int r2,int c2) {
        int ans=0;
        for (int i=r1;i<=r2;i++) {
            ans+=st.QueryRange(1,1,m*n,pno(i,c1),pno(i,c2));
        }
        return ans;
    }
};

File: vol1/ch12/LeetCode308.py
class SegTree:             #线段树类
    def __init__(self,a):
        n=len(a)
        self.a=a
        self.ST=[0]*(4*n)
        self.Create(1,1,n)
    def Create(self,root,L,R):
        if L==R:self.ST[root]=self.a[L-1]
        else:
            mid=(L+R)>>1
            self.Create(2*root,L,mid)
            self.Create(2*root+1,mid+1,R)
            self.ST[root]=self.ST[2*root]+self.ST[2*root+1]
    def UpdateOne(self,root,L,R,i,x):          #单点修改:a[i]+=x
        if L==R:self.ST[root]=x                                            #单点
        else:                                          #非单点
            mid=(L+R)>>1
            if i<=mid:self.UpdateOne(2*root,L,mid,i,x);	     #在左子树中修改
            else: self.UpdateOne(2*root+1,mid+1,R,i,x); #在右子树中修改
            self.ST[root]=self.ST[2*root]+self.ST[2*root+1];       #向上更新分支结点值
    def QueryRange(self,root,L,R,i,j):          #查询区间[i,j]
        if L>=i and R<=j:return self.ST[root]        #当前结点[L,R]包含在[i,j]中
        else:
            mid=(L+R)>>1                           #其他情况
            lsum,rsum=0,0
            if i<=mid:                                      #若[i,j]与左孩子有交集
                lsum=self.QueryRange(2*root,L,mid,i,j)
            if mid<j:                                           #若[i,j]与右孩子有交集
                rsum=self.QueryRange(2*root+1,mid+1,R,i,j)
            return lsum+rsum;
class NumMatrix:
    def __init__(self, matrix: List[List[int]]):
        self.m,self.n=len(matrix),len(matrix[0])
        a=[]
        for i in range(0,self.m):
            for j in range(0,self.n):
                a.append(matrix[i][j])
        self.st=SegTree(a)
    def pno(self,i,j):      #二维转换为一维
        return i*self.n+j+1
    def update(self,r:int,c:int,val: int) -> None:
        i=self.pno(r,c)
        self.st.UpdateOne(1,1,self.m*self.n,i,val)
    def sumRegion(self,r1:int,c1:int,r2:int,c2:int) -> int:
        ans=0
        for i in range(r1,r2+1):
            ans+=self.st.QueryRange(1,1,self.m*self.n,self.pno(i,c1),self.pno(i,c2))
        return ans
